### ISSUE TITLE
cumprod backward for zero-length tensors

### ISSUE THREAD

**Issue by leedtan** [pytorch/pytorch#31782]
## 🐛 Bug
Calling cumsum on a zero-dimension tensor raises an error in backward.

This bug is very similar to issue https://github.com/pytorch/pytorch/issues/31515, but for cumprod as opposed to cumsum.

## To Reproduce

import torch

x = torch.zeros(2, 0, requires_grad=True)
xx = x.cumprod(dim=-1).sum()
xx.backward()

import torch

x = torch.zeros(2, 0, requires_grad=True)
xx = x.cumsum(dim=-1).sum()
xx.backward()

## Expected behavior

I would expect no error message. If there is no gradient because there were no values to back-propagate it could optionally provide a warning. 

## Environment

Collecting environment information...
PyTorch version: 1.4.0a0+2aa5aa6
Is debug build: No
CUDA used to build PyTorch: 9.2

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
CMake version: version 3.14.0

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration: 
GPU 0: Tesla M40
GPU 1: Tesla M40

Nvidia driver version: 396.69
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] torch==1.4.0a0+2aa5aa6
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.0.15           py38ha843d7b_0  
[conda] mkl_random                1.1.0            py38h962f231_0  
[conda] torch                     1.4.0a0+2aa5aa6           dev_0    <develop>

**Comment by mrshenli** [comment]
Sel-assigned to reserve for bootcamp

**Comment by mrshenli** [comment]
fixed by #32070
